rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,934 @@
+It will read the unauthenticated `/v1/webapi/find` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified update plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
+
+Source code for the updater will live in the main Teleport repository, with the updater binary built from `tools/teleport-update`.
+
+### Installation
+
+```shell
+$ apt-get install teleport-ent-updater
+$ teleport-update enable --proxy example.teleport.sh
+
+# if not enabled already, configure teleport and:
+$ systemctl enable teleport
+```
+
+For grouped updates, a group identifier may be configured:
+```shell
+$ teleport-update enable --proxy example.teleport.sh --group staging
+```
+
+For air-gapped Teleport installs, the agent may be configured with a custom tarball path template:
+```shell
+$ teleport-update enable --proxy example.teleport.sh --template 'https://example.com/teleport-{{ .Edition }}-{{ .Version }}-{{ .Arch }}.tgz'
+```
+(Checksum will use template path + `.sha256`)
+
+### Filesystem
+
+```
+$ tree /var/lib/teleport
+/var/lib/teleport
+└── versions
+   ├── 15.0.0
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-update
+   │  │  └── teleport
+   │  ├── etc
+   │  │  └── systemd
+   │  │     └── teleport.service
+   │  └── backup
+   │     ├── sqlite.db
+   │     └── backup.yaml
+   ├── 15.1.1
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-update
+   │  │  └── teleport
+   │  └── etc
+   │     └── systemd
+   │        └── teleport.service
+   └── updates.yaml
+$ ls -l /usr/local/bin/tsh
+/usr/local/bin/tsh -> /var/lib/teleport/versions/15.0.0/bin/tsh
+$ ls -l /usr/local/bin/tbot
+/usr/local/bin/tbot -> /var/lib/teleport/versions/15.0.0/bin/tbot
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport-update
+/usr/local/bin/teleport-update -> /var/lib/teleport/versions/15.0.0/bin/teleport-update
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+kind: agent_versions
+spec:
+  # proxy specifies the Teleport proxy address to retrieve the agent version and update configuration from.
+  proxy: mytenant.teleport.sh
+  # group specifies the update group
+  group: staging
+  # enabled specifies whether auto-updates are enabled, i.e., whether teleport-update update is allowed to update the agent.
+  enabled: true
+  # active_version specifies the active (symlinked) deployment of the telepport agent.
+  active_version: 15.1.1
+```
+
+backup.yaml:
```

Comment:
What is this supposed to do? I'm not sure if I missed something above or just haven't got to it yet, but I'm finding it hard to comprehend with just a snippet of YAML and no other commentary.

Additionally, when I see `kind: db_backup` nothing about that name suggests that this is in any way related to agent auto upgrades. Is that intended?

Reply:
Thanks, I'll make this more explicit in the section below.

It's metadata about the sqlite DB backup, referred to here: https://github.com/gravitational/teleport/pull/40190/files#diff-4fb277bd3d884c7b045e7efad61111ac4e7957c063e0c0c9281f4e74fd72997fR571

(Users don't need to worry about this file -- it's an implementation detail.)

Reply:
> nothing about that name suggests that this is in any way related to agent auto upgrades. Is that intended?

This file is not a Teleport resource (see https://github.com/gravitational/teleport/pull/40190#discussion_r1553934860). It's machine-local configuration storage for individual Teleport backups. That said, the metadata is descriptive of any backup of Teleport's sqlite DB, so it could describe non-agent data.